### PR TITLE
Update Jax(LLVM, MLIR) and Enzyme

### PR DESCRIPTION
**Pull request by rmoyard**
**Description of the Change:**
NullOp -> ZeroOp
https://github.com/llvm/llvm-project/pull/67183

CodeGenOpt -> CodeGenOptLevel
https://github.com/llvm/llvm-project/pull/66295/files

Converters are const
https://github.com/llvm/llvm-project/commit/ce254598b73b119c9463f5b7f4131559e276e844

Remove typed pointer
https://github.com/llvm/llvm-project/pull/71285

thlo gml_st on ice
https://github.com/tensorflow/mlir-hlo/commit/7077cec01a77db8f1f3e2537c0852281a89ad4d2

ByteCode is a new interface
https://github.com/llvm/llvm-project/commit/837d1ce0dc8eec5b17255291b3462e6296cb369b

getArgsOperandMutable was added CallOpInterface
https://github.com/llvm/llvm-project/commit/d790a217a746ff63190d30669674fce1bc0a4723

Remove redundant memref.tensor_store op
https://github.com/llvm/llvm-project/pull/71010/files

Lowering of memref.realloc was extracted to its own pass
https://reviews.llvm.org/D159430

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: mlir/lib/Catalyst/Transforms/catalyst_to_llvm.cpp [hunk 1]
```diff
@@ -181,7 +181,7 @@ Value EncodeOpaqueMemRef(Location loc, PatternRewriter &rewriter, MemRefType mem
     Type voidPtrType = LLVM::LLVMPointerType::get(ctx);
     Value c1 = rewriter.create<LLVM::ConstantOp>(loc, rewriter.getI64IntegerAttr(1));
     Value memrefPtr =
-        rewriter.create<LLVM::AllocaOp>(loc, LLVM::LLVMPointerType::get(llvmMemrefType), c1);
+        rewriter.create<LLVM::AllocaOp>(loc, LLVM::LLVMPointerType::get(rewriter.getContext()), c1);
```
Comment (rmoyard) [inline mlir/lib/Catalyst/Transforms/catalyst_to_llvm.cpp:184]: Update

File: mlir/lib/Quantum/Transforms/quantum_to_llvm.cpp [hunk 2]
```diff
@@ -47,12 +47,12 @@ struct QIRTypeConverter : public LLVMTypeConverter {
   private:
     Type convertQubitType(Type mlirType)
     {
-        return LLVM::LLVMPointerType::get(LLVM::LLVMStructType::getOpaque("Qubit", &getContext()));
+        return LLVM::LLVMPointerType::get(&getContext());
```
Comment (rmoyard) [inline mlir/lib/Quantum/Transforms/quantum_to_llvm.cpp:50]: Not sure about this file

File: mlir/lib/Quantum/Transforms/emit_catalyst_pyface.cpp [hunk 3]
```diff
@@ -154,7 +154,8 @@ void wrapResultsAndArgsInTwoStructs(LLVM::LLVMFuncOp op, PatternRewriter &rewrit
 
     if (hasInputs) {
         Value arg = wrapperFuncOp.getArgument(1);
-        Value structOfMemrefs = rewriter.create<LLVM::LoadOp>(loc, arg);
+        auto argType = wrapperFuncType.getParamType(1);
```
Comment (rmoyard) [inline mlir/lib/Quantum/Transforms/emit_catalyst_pyface.cpp:157]: Not sure about that given the next file

Comment (erick-xanadu) [reply mlir/lib/Quantum/Transforms/emit_catalyst_pyface.cpp:157]: I think this should be good.

**Comment by rmoyard** [comment]
[sc-53151]

File: mlir/lib/Quantum/Transforms/emit_catalyst_pyface.cpp [hunk 4]
```diff
@@ -131,6 +121,17 @@ void wrapResultsAndArgsInTwoStructs(LLVM::LLVMFuncOp op, PatternRewriter &rewrit
     bool hasInputs = functionHasInputs(callee);
 
     LLVM::LLVMFunctionType functionType = op.getFunctionType();
+    Type ptrType = LLVM::LLVMPointerType::get(rewriter.getContext());
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply mlir/lib/Quantum/Transforms/emit_catalyst_pyface.cpp:124]:
```
!llvm.ptr<struct<(ptr<struct<(ptr<f64>, ptr<f64>, i64)>>, ptr<struct<(ptr<f64>, ptr<f64>, i64)>>)>>
```
things like that appear in the IR

[the remark this answers is not in the document]
Comment (erick-xanadu) [reply mlir/lib/Quantum/Transforms/emit_catalyst_pyface.cpp:124]: Can you take a look at the next file? One ending in `.ll` not in `.mlir`?

File: mlir/test/Catalyst/BufferizationTest.mlir [hunk 5]
```diff
@@ -1,6 +1,6 @@
 // Copyright 2023 Xanadu Quantum Technologies Inc.
 
-// Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (dime10) [inline mlir/test/Catalyst/BufferizationTest.mlir:3]: This guy lost his brother :(

File: mlir/lib/Quantum/Transforms/emit_catalyst_pyface.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (dime10) [reply mlir/lib/Quantum/Transforms/emit_catalyst_pyface.cpp:124]: For this specific file @erick-xanadu would know best if the patch makes sense, but the general approach in the PR seems correct to me 👍  We are just moving the "loss" of type information one level up, from the LLVM IR into the LLVM dialect. The only downside I see is tests might be a bit harder if they were specifically testing the type information.

Now looking at the code in a lot of places it can seem a bit redundant, like writing `charPtr = LLVM::LLVMPointerType()` and then writing it again for a different pointer, even though all pointers are the same, but it doesn't do any harm either.

File: mlir/lib/Quantum/Transforms/quantum_to_llvm.cpp [hunk 6]
```diff
@@ -47,12 +47,12 @@ struct QIRTypeConverter : public LLVMTypeConverter {
   private:
     Type convertQubitType(Type mlirType)
     {
-        return LLVM::LLVMPointerType::get(LLVM::LLVMStructType::getOpaque("Qubit", &getContext()));
+        return LLVM::LLVMPointerType::get(&getContext()); // LLVM::LLVMStructType::getOpaque("Qubit", &getContext());
```
[the remark this answers is not in the document]
Comment (dime10) [reply mlir/lib/Quantum/Transforms/quantum_to_llvm.cpp:50]: We should only ever be loading sized types. For example a Qubit should be loaded as follows:

before
```ll
  %1 = call %Array* @__quantum__rt__qubit_allocate_array(i64 2)
  %2 = call i8* @__quantum__rt__array_get_element_ptr_1d(%Array* %1, i64 0)
  %3 = bitcast i8* %2 to %Qubit**
  %4 = load %Qubit*, %Qubit** %3, align 8
```
now
```ll
  %1 = call ptr @__quantum__rt__qubit_allocate_array(i64 2)
  %2 = call ptr @__quantum__rt__array_get_element_ptr_1d(ptr %1, i64 0)
  ; bitcast no longer needed
  %3 = load ptr, ptr %2, align 8
```

So here the type that is loaded is `ptr`. Let me know if that makes sense.

[the remark this answers is not in the document]
Comment (rmoyard) [reply mlir/lib/Quantum/Transforms/quantum_to_llvm.cpp:50]: Yes I think that is happening right now

File: frontend/test/pytest/test_jax_dynamic_api.py [hunk 7]
```diff
@@ -347,6 +347,7 @@ def f():
         qjit(f)
 
 
+@pytest.mark.xfail
 def test_shapes_type_conversion():
```
Comment (rmoyard) [inline frontend/test/pytest/test_jax_dynamic_api.py:351]: @dime10 @grwlf It is failing because of this PR https://github.com/google/jax/commit/a1ee8c174386124973f7d898164d487cd3e0e579 in Jax. Could let me know if you see any solution?

File: mlir/lib/Gradient/Transforms/GradMethods/PS_QuantumGradient.cpp [hunk 8]
```diff
@@ -156,7 +156,11 @@ static void storePartialDerivative(PatternRewriter &rewriter, Location loc,
             Value gradientSubview =
                 rewriter.create<memref::SubViewOp>(loc, resultType, gradientBuffer, dynOffsets,
                                                    dynSizes, dynStrides, offsets, sizes, strides);
-            rewriter.create<memref::TensorStoreOp>(loc, derivative, gradientSubview);
+            derivative.getType().dump();
```
Comment (rmoyard) [inline mlir/lib/Gradient/Transforms/GradMethods/PS_QuantumGradient.cpp:159]: Remove

File: frontend/test/pytest/test_jax_dynamic_api.py (hunk 7 above)
Comment (dime10) [reply frontend/test/pytest/test_jax_dynamic_api.py:351]: As discussed in the meeting, we can change this test to verify floats _raise_ an error, although technically this is something JAX should be doing, so I think it would fine removing it altogether as well.

**Comment by rmoyard** [comment]
Blocked by https://github.com/EnzymeAD/Enzyme/pull/1608

File: frontend/test/pytest/test_jax_dynamic_api.py (hunk 7 above)
Comment (rmoyard) [reply frontend/test/pytest/test_jax_dynamic_api.py:351]: Test removed

File: mlir/lib/Gradient/Transforms/GradMethods/PS_QuantumGradient.cpp [hunk 9]
```diff
@@ -156,7 +156,11 @@ static void storePartialDerivative(PatternRewriter &rewriter, Location loc,
             Value gradientSubview =
                 rewriter.create<memref::SubViewOp>(loc, resultType, gradientBuffer, dynOffsets,
                                                    dynSizes, dynStrides, offsets, sizes, strides);
-            rewriter.create<memref::TensorStoreOp>(loc, derivative, gradientSubview);
+
+            gradientSubview.getType().dump();
```
Comment (erick-xanadu) [inline mlir/lib/Gradient/Transforms/GradMethods/PS_QuantumGradient.cpp:160]:
```suggestion
```

File: .dep-versions [hunk 10]
```diff
@@ -1,5 +1,4 @@
-# Update the version check in catalyst.__init__ when changing the JAX version.
-jax=0.4.14
-mhlo=00be4a6ce2c4d464e07d10eae51918a86f8df7b4
-llvm=4706251a3186c34da0ee8fd894f7e6b095da8fdc
-enzyme=8d22ed1b8c424a061ed9d6d0baf0cc0d2d6842e2
+jax=0.4.23
```
Comment (dime10) [inline .dep-versions:1]: We should probably keep this comment for the future:
```suggestion
# Always update the version check in catalyst.__init__ when changing the JAX version.
jax=0.4.23
```

File: frontend/catalyst/compiler.py [hunk 11]
```diff
@@ -128,7 +128,6 @@ def run_writing_command(command: List[str], compile_options: Optional[CompileOpt
         "func.func(hlo-legalize-to-linalg)",
         "func.func(mhlo-legalize-to-std)",
         "convert-to-signless",
-        "func.func(scalarize)",
```
Comment (dime10) [inline frontend/catalyst/compiler.py:131]: @erick-xanadu I remember you found that either of two passes was providing similar speedups, one at the hlo level and one at the linalg level. Can we replace this pass with the linalg one (since it no longer exists)?

File: frontend/catalyst/compiler.py [hunk 12]
```diff
@@ -179,6 +178,7 @@ def run_writing_command(command: List[str], compile_options: Optional[CompileOpt
 MLIR_TO_LLVM_PASS = (
     "MLIRToLLVMDialect",
     [
+        "expand-realloc",
```
Comment (dime10) [inline frontend/catalyst/compiler.py:181]: Small comment why this pass is necessary will be nice, since it's hard to see why pipelines are the way they are.

File: frontend/test/lit/test_aot_compilation.py [hunk 13]
```diff
@@ -426,8 +426,6 @@ def multiple_stablehlo_custom_call(A: ShapedArray([2, 2], jax.numpy.float64)):
 
     # CHECK: func.func private @eigh
     # CHECK: stablehlo.custom_call @lapack_dsyevd
-    # CHECK: func.func private @eigh_1
-    # CHECK: stablehlo.custom_call @lapack_dsyevd
```
Comment (dime10) [inline frontend/test/lit/test_aot_compilation.py:430]: This is a bugfix right?

File: mlir/lib/Gradient/Transforms/ConversionPatterns.cpp [hunk 14]
```diff
@@ -160,8 +160,8 @@ static constexpr const char *enzyme_dupnoneed_key = "enzyme_dupnoneed";
 /// functions where MemRefs are passed via wrapped pointers (!llvm.ptr<struct(ptr, ptr, i64, ...)>)
 /// rather than having their fields unpacked. This function automatically transforms MemRef
 /// arguments of a function to wrapped pointers.
-void wrapMemRefArgs(func::FuncOp func, LLVMTypeConverter &typeConverter, PatternRewriter &rewriter,
-                    Location loc, bool volatileArgs = false)
+void wrapMemRefArgs(func::FuncOp func, const LLVMTypeConverter *typeConverter,
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply mlir/lib/Gradient/Transforms/ConversionPatterns.cpp:163]: Because they change the type of the converter to be constant, I think we had lvalue/rvalue

File: mlir/test/Catalyst/BufferizationTest.mlir [hunk 15]
```diff
@@ -30,7 +30,7 @@ func.func @dbprint_val(%arg0: tensor<?xf64>) {
 
 func.func @dbprint_memref(%arg0: tensor<?xf64>) {
 
-    // CHECK: "catalyst.print"(%0) {print_descriptor} : (memref<?xf64>) -> ()
+    // CHECK: "catalyst.print"(%0) <{print_descriptor}> : (memref<?xf64>) -> ()
```
Comment (dime10) [inline mlir/test/Catalyst/BufferizationTest.mlir:33]: Is this the attribute dictionary that is now printed like this?

File: mlir/test/Quantum/WrapperTest.mlir
Comment (dime10) [inline mlir/test/Quantum/WrapperTest.mlir:1]: So much cleaner these tests 😍

**Review by dime10 — APPROVED** [review]
Thanks @rmoyard for this long awaited update 💯

File: .dep-versions (hunk 10 above)
Comment (rmoyard) [reply .dep-versions:1]: Good catch!

File: mlir/test/Catalyst/BufferizationTest.mlir (hunk 15 above)
Comment (rmoyard) [reply mlir/test/Catalyst/BufferizationTest.mlir:33]: Honestly I don't know, I did not have to change anything expect that in the test. I could not find where the change comes from

File: frontend/test/lit/test_aot_compilation.py (hunk 13 above)
Comment (rmoyard) [reply frontend/test/lit/test_aot_compilation.py:430]: I think it is an lowering update from jax -> stable hlo

Comment (dime10) [reply frontend/test/lit/test_aot_compilation.py:430]: Oh ok, nvm 👍

File: frontend/catalyst/compiler.py (hunk 12 above)
Comment (rmoyard) [reply frontend/catalyst/compiler.py:181]: https://reviews.llvm.org/D159430 This was part of the memref lowering and it was extracted to make lowering simpler

Comment (dime10) [reply frontend/catalyst/compiler.py:181]: Right I meant adding it as a comment into the file
